src/TesApi.Web/TesEventMessage.cs

```diff
@@ -0,0 +1,178 @@
+            return (result.EntityId, GetCompletedBatchState(result));
+        }
+
+        /// <summary>
+        /// Marks this event message processed.
+        /// </summary>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        public async Task MarkMessageProcessed(CancellationToken cancellationToken)
+        {
+            await _azureProxy.SetBlobTags(
+                _uri,
+                Tags.Append(new(ProcessedTag, DateTime.UtcNow.ToString("O")))
+                    .ToDictionary(pair => pair.Key, pair => pair.Value),
+                cancellationToken);
+        }
+
+        private /*static*/ AzureBatchTaskState GetCompletedBatchState(Tes.Runner.Events.EventMessage task)
+        {
+            return (task.Name ?? Event) switch
+            {
+                Tes.Runner.Events.EventsPublisher.TaskCompletionEvent => string.IsNullOrWhiteSpace(task.EventData["errorMessage"])
+
+                    ? new(
+                        AzureBatchTaskState.TaskState.CompletedSuccessfully,
+                        BatchTaskStartTime: task.Created - TimeSpan.Parse(task.EventData["duration"]),
+                        BatchTaskEndTime: task.Created/*,
+                            BatchTaskExitCode: 0*/)
+
+                    : new(
+                        AzureBatchTaskState.TaskState.CompletedWithErrors,
+                        Failure: new("ExecutorError",
+                        Enumerable.Empty<string>()
+                            .Append(task.EventData["errorMessage"])),
+                        BatchTaskStartTime: task.Created - TimeSpan.Parse(task.EventData["duration"]),
+                        BatchTaskEndTime: task.Created/*,
+                            BatchTaskExitCode: 0*/),
+
+                // TODO: the rest
+                _ => new(AzureBatchTaskState.TaskState.NodePreempted), //throw new System.Diagnostics.UnreachableException(),
+            };
+        }
+    }
+
+    // TODO: Consider moving this class's implementation to Startup
+    /// <summary>
+    /// Factory to create TesEventMessage instances.
+    /// </summary>
+    /// <remarks>This can be a singleton in DI.</remarks>
+    public sealed class BatchTesEventMessageFactory
+    {
+        private readonly IServiceProvider _serviceProvider;
+
+        /// <summary>
+        /// Constructor for <see cref="BatchPoolFactory"/>.
+        /// </summary>
+        /// <param name="serviceProvider">A service object.</param>
+        public BatchTesEventMessageFactory(IServiceProvider serviceProvider) => _serviceProvider = serviceProvider;
+
+        /// <summary>
+        /// Creates a new <see cref="TesEventMessage"/>.
+        /// </summary>
+        /// <param name="blobAbsoluteUri"></param>
+        /// <param name="tags"></param>
+        /// <param name="event"></param>
+        /// <returns></returns>
+        public TesEventMessage CreateNew(Uri blobAbsoluteUri, IDictionary<string, string> tags, string @event)
```

Comment:
This class exists because the tags, event, and URI are part of the constructor, and they change per instance. Instead, I would pass them to the method that uses them, simplify the constructor, and register the type at start up directly in DI.